### PR TITLE
Add more advection scheme tests; fix order adapting with tracer-specific schemes; fix WENOVectorInvariant

### LINKED ISSUE #3863: Cannot build `HydrostaticFreeSurfaceModel` with different advection schemes for different tracers
There's a bug in `adapt_advection_order` (or the constructor for `HydrostaticFreeSurfaceModel`, since this code:

```julia
using Oceananigans
grid = RectilinearGrid(size=1, x=(0, 1), topology=(Periodic, Flat, Flat))
tracer_advection = (b=nothing, c=WENO())
model = HydrostaticFreeSurfaceModel(; grid, tracer_advection)
```

fails with

```
ERROR: LoadError: MethodError: no method matching adapt_advection_order(::@NamedTuple{…}, ::Int64, ::RectilinearGrid{…})

Closest candidates are:
  adapt_advection_order(::Nothing, ::Int64, ::Oceananigans.Grids.AbstractGrid)
   @ Oceananigans ~/Projects/Oceananigans.jl/src/Advection/adapt_advection_order.jl:60
  adapt_advection_order(::UpwindBiased{B}, ::Int64, ::Oceananigans.Grids.AbstractGrid) where B
   @ Oceananigans ~/Projects/Oceananigans.jl/src/Advection/adapt_advection_order.jl:74
  adapt_advection_order(::WENO{B, FT, XT, YT, ZT}, ::Int64, ::Oceananigans.Grids.AbstractGrid) where {B, FT, XT, YT, ZT}
   @ Oceananigans ~/Projects/Oceananigans.jl/src/Advection/adapt_advection_order.jl:92
```

### LINKED ISSUE #3865: Bug in WENOVectorInvariant constructor
`required_halo_size` does not exist:

https://github.com/CliMA/Oceananigans.jl/blob/03b8acf4f378eeefdb5e79ceeafcf29fa711e94c/src/Advection/vector_invariant_advection.jl#L215

I guess it's not tested.

### PR DESCRIPTION
**Pull request by glwagner**
Closes #3863 (+ fixes FluxFormAdvection docstring)
Closes #3865

### CODE REVIEW

File: test/test_hydrostatic_free_surface_models.jl [hunk 1]
```diff
@@ -210,24 +211,34 @@ topos_3d = ((Periodic, Periodic, Bounded),
...
 
         momentum_advection = VectorInvariant()
         @testset "Time-stepping HydrostaticFreeSurfaceModels [$arch, $(typeof(momentum_advection))]" begin
             @info "  Testing time-stepping HydrostaticFreeSurfaceModels [$arch, $(typeof(momentum_advection))]..."
-            @test time_step_hydrostatic_model_works(lat_lon_sector_grid, momentum_advection=momentum_advection)
+            @test time_step_hydrostatic_model_works(lat_lon_sector_grid; momentum_advection)
         end
 
+        for tracer_advection in [WENO(),
+                                 FluxFormAdvection(WENO(), WENO(), Centered())
```
Comment (navidcy) [inline test/test_hydrostatic_free_surface_models.jl:233]: minor comment: is there a reason why the specific addiction schemes were chosen? if so perhaps a comment is good?

**Review by navidcy — APPROVED** [review]
lgtm!

File: test/test_hydrostatic_free_surface_models.jl (hunk 1 above)
Comment (glwagner) [reply test/test_hydrostatic_free_surface_models.jl:233]: More cases can be added. This just spans three significantly different possibilities. Only the 3rd is broken currently

Comment (glwagner) [reply test/test_hydrostatic_free_surface_models.jl:233]: All the other tests are the same btw. They are not exhaustive

Comment (navidcy) [reply test/test_hydrostatic_free_surface_models.jl]: I didn't mean to imply we need to add more test. I was just wondering if there was a reason z was different than the x, y (and if so, perhaps a comment would be appropriate in that case). But if it's just randomness then no worries!

**Comment by glwagner** [comment]
@navidcy if you had a problem with FluxFormAdvection docstring you may want to look at WENOVectorInvariant...

**Review by simone-silvestri — APPROVED** [review]

**Comment by navidcy** [comment]
> @navidcy if you had a problem with FluxFormAdvection docstring you may want to look at WENOVectorInvariant...

I didn't have a problem per se. But it doesn't hurt to improve.

**Pull request merged**